app/src/main/java/org/wikipedia/edit/SyntaxHighlightableEditText.kt

```diff
@@ -0,0 +1,208 @@
+package org.wikipedia.edit
+
+import android.annotation.SuppressLint
+import android.content.Context
+import android.graphics.Canvas
+import android.graphics.Paint
+import android.graphics.Rect
+import android.os.Build
+import android.os.SystemClock
+import android.text.Editable
+import android.text.InputType
+import android.text.Layout
+import android.text.TextPaint
+import android.util.AttributeSet
+import android.view.KeyEvent
+import android.view.View
+import android.view.inputmethod.EditorInfo
+import android.view.inputmethod.InputConnection
+import android.widget.EditText
+import androidx.core.view.ViewCompat
+import org.wikipedia.R
+import org.wikipedia.util.DimenUtil
+import org.wikipedia.util.ResourceUtil
+
+/**
+ * Notice that this view inherits from the platform EditText class, instead of AppCompatEditText.
+ * This is because AppCompatEditText actually has very poor performance when displaying large
+ * amounts of text. Inheriting directly from the base EditText provides a vast improvement in
+ * performance, but it means we need to keep an eye on the source code of AppCompatEditText, in
+ * case we need to copy over any useful compatibility logic.
+ */
+@SuppressLint("AppCompatCustomView")
+open class SyntaxHighlightableEditText : EditText {
+
+    private var prevLineCount = -1
+    private val lineNumberPaint = TextPaint()
+    private val lineNumberBackgroundPaint = Paint()
+    private val isRtl: Boolean
+    private val paddingWithoutLineNumbers = DimenUtil.roundedDpToPx(8f)
+    private val paddingWithLineNumbers = DimenUtil.roundedDpToPx(36f)
+    private val lineNumberGapWidth = DimenUtil.roundedDpToPx(8f)
+    private var allowScrollToCursor = true
+
+    lateinit var scrollView: View
+    private lateinit var actualLineFromRenderedLine: IntArray
+
+    var inputConnection: InputConnection? = null
+
+    var showLineNumbers = false
+        set(value) {
+            field = value
+            applyPaddingForLineNumbers()
+        }
+
+    constructor(context: Context) : super(context)
+    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs)
+    constructor(context: Context, attrs: AttributeSet?, defStyle: Int) : super(context, attrs, defStyle)
+
+    init {
+        // The MIME type(s) need to be set for onReceiveContent() to be called.
+        ViewCompat.setOnReceiveContentListener(this, arrayOf("text/*"), null)
```

Comment:
I notice the `this` has some warning message. Should we make the class `final`?

Reply:
yep